### ISSUE TITLE
[exporter/otlphttp] 500 should be retryable or not

### ISSUE THREAD

**Issue by yutingcaicyt** [open-telemetry/opentelemetry-specification#3314]
**What are you trying to achieve?**
For otlphttp exporter, add 500 to the retriable code. 
What did you expect to see?
The retriable code is set to be [429, 500, 502, 503, 504]
**Additional context.**
After 0.71.0, the retry code for otlphttp exporter is set to be [429/502/503/504].
In my situation, the servers dealing with data from otel-collector are deployed on a batch of VM. When these servers are redeployed, the otel-collector may receive some 500 errors. This is because there may be still some requests being processed when the resource of the server is destroyed. 
In the situation mentioned above, the 500 is not a permanent error. It will be recovered after the server is restarted.
500 means an internal server error, it may happen in the case mentioned above or any other coding error. I think no matter what the situation, the error 500 is expected to be resolved. Once it's fixed, it will make sense to retry the request. So I wonder if 500 can be added to the retriable code.

**Comment by jack-berg** [comment]
I disagree that 500 is a retriable in the general case. While there may be cases (like yours) where its a temporary error, 500 is a general catch all error and more often than not, represents some sort of bug / programming error that will not resolve with retries. In cases like yours where retrying a 500 would resolve with time, there's probably a more appropriate status code to use, e.g. 503 in your case.

**Comment by yutingcaicyt** [comment]
> I disagree that 500 is a retriable in the general case. While there may be cases (like yours) where its a temporary error, 500 is a general catch all error and more often than not, represents some sort of bug / programming error that will not resolve with retries. In cases like yours where retrying a 500 would resolve with time, there's probably a more appropriate status code to use, e.g. 503 in your case.

I agree that retrying the request with 500 error can't resolve programming error. But what exactly is a retriable error? In fact, there is no guarantee that retrying will work even with code 429, 502, 503, 504 in the retry cycle.  We retry the error 429, 502, 503, 504 because the request may be processed successfully with a certain probability in the retry cycle. If we think about it from this point of view, requests with 500 are also likely to retry successfully(of course it may not work)。 Then it makes sense to put 500 into the retriable code. Of course we have to control the times and frequency of retries to avoid infinite reties.

**Comment by kgoralski** [comment]
Hey, what if the sender will receive just 500 Internal Server Errors? It will not retry on it but I think it should? Having strict number of status codes to retry is fine when you can have a strong contract between client and server. But for generic exporter it's hard to expect something. Better to have a metric for this and allow "admins" to react on this.

In the case of using daemonsets -> gateway and gateway responding with a high memory limit. It will respond with:

```shell
Exporting failed. The error is not retryable. Dropping data.
Permanent error: error exporting items, request to http://gateway.otel.svc.cluster.local:4318/v1/logs 
responded with HTTP Status Code 500, Message=data dropped due to high memory usage, Details=[]
```
And agents will simply drop data in such case.
I think it looks like a retriable error? For a agent and gateway case it would be nice to define separate 5xx code that can be retried.

**Comment by kgoralski** [comment]
Any feedback on this? @newly12 any ideas? 
Overall I think it should be flexible and let user decide on what status codes to retry.

**Comment by jack-berg** [comment]
> I think if the sender will receive just 500 Internal Server Error it should be retried. It's almost always considered as temporary error.

There is not consensus on that. 

> In the case of using daemonsets -> gateway and gateway responding with a high memory limit. It will respond with:
> Exporting failed. The error is not retryable. Dropping data.
Permanent error: error exporting items, request to http://gateway.otel.svc.cluster.local:4318/v1/logs 
responded with HTTP Status Code 500, Message=data dropped due to high memory usage, Details=[]

The collector should consider returning a more appropriate status code in this case. 

> In various API docs you will find 500 error as something retryable, for example PayPal

OTLP is a specification and as part of that specification declares certain behavior around certain status codes. Servers acting as OTLP receivers need to be aware of the specification and the semantics that different status codes carry. If a OTLP receiver returns a 500 in a retryable situation, then it's not conforming to the specification.

Sometimes there are load balancers or other intermediate servers involved which aren't aware of the OTLP specification and can't easily be programmed to adjust their behavior to conform to it. But in my experience these components return status codes like 502 or 503, not 500, when not able to communicate with the downstream target, and are therefore not a problem.  

Also, the [OTLP specification is stable](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/protocol/otlp.md#opentelemetry-protocol-specification). Changing the retryable status codes would be considered a breaking change and is not allowed. Maybe we could adjust the status codes to be configurable, but I don't think we should.

**Comment by kgoralski** [comment]
> The collector should consider returning a more appropriate status code in this case.

I agree. So in that particular case, is it's something ok to implement it in OTEL stack to fix that problem? And it should be not against the specification? 

> OTLP is a specification and as part of that specification declares certain behavior around certain status codes. Servers acting as OTLP receivers need to be aware of the specification and the semantics that different status codes carry. If a OTLP receiver returns a 500 in a retryable situation, then it's not conforming to the specification.

I agree but the backend for the exporter can be nearly anything, not necessarily something that can be changed/controlled.  By "Servers acting as OTLP receivers", do you mean HTTP/gRPC servers or also things like Kafka, Kinesis/Prometheus/Datadog etc. ? 

> Sometimes there are load balancers or other intermediate servers involved which aren't aware of the OTLP specification and can't easily be programmed to adjust their behavior to conform to it.

The current state will force people to work around the problem using load balancers and rewriting the status codes

> But in my experience these components return status codes like 502 or 503, not 500

That's true, but. Usually, when the backend is talking to another backend, like Database and there will be a temporary DB failure it will rather throw 500, maybe eventually somebody will change it to 503, but can be a rare thing. Eventually, such a backend will be removed from LoadBalancer failing readiness check but it can take some time. In our case 500 = immediately data drop.

**Comment by timtofan** [comment]
> In our case 500 = immediately data drop.
Same for us.

Out of curiosity, what is the logic in treating 500 as non-retryable?
As in, how is it useful in the context of otlp?
